BUILD_TOPOLOGY.toml

```diff
@@ -9,6 +9,41 @@
 version = "1.0"
 description = "TheRock artifact-based build topology"
 
+# ==============================================================================
+# SOURCE SETS - Submodule groupings for partial checkouts
+# ==============================================================================
+# Each source_set defines submodules needed for a category of builds.
+# Currently just lists submodules; can be extended later with sparse_checkout specs.
+# Names are chosen to align with fetch_sources.py flags where possible.
+
+[source_sets.base]
+description = "Base infrastructure submodules"
+submodules = ["half", "rocm-cmake", "amdsmi"]
+
+[source_sets.compilers]
+description = "Compiler toolchain submodules"
+submodules = ["llvm-project", "HIPIFY", "spirv-llvm-translator"]
+
+[source_sets.rocm-systems]
+description = "ROCm systems monorepo (runtime, profiler, etc.)"
+submodules = ["rocm-systems"]
```

Comment:
FYI @ammallya for submodule update automation, we can try to track these "source sets" (https://github.com/ROCm/TheRock/pull/2252#discussion_r2550601378)

(We found there that dependabot does not support grouping git submodule updates like "always update rccl and rccl-tests together" though)